lib/brod_ex/client.ex

```diff
@@ -1,33 +1,75 @@
 defmodule BrodEx.Client do
-  @moduledoc false
+  @moduledoc """
+  Defines a client to interact with a Kafka cluster.
 
-  defmacro __using__(opts \\ []) do
-    quote bind_quoted: [opts: opts] do
-      otp_app = Keyword.fetch!(opts, :otp_app)
+  Similar to an Ecto Repo, you define your own module and configure connection
+  and client behaviour in your `config/config.exs.`
 
-      @otp_app otp_app
+  When used the client expects the `:otp_app` as option. The `:otp_app` should point
+  to an OTP application that has the Kafka client configuration. For example, the client:
 
-      @default_brod_client :default_brod_client
+      defmodule MyApp.Kafka do
+        use BrodEx.Client, otp_app: :my_app
+      end
+
+  Can be configured with:
+
+      config :your_app, YourApp.Kafka,
+        endpoints: "localhost:9092"
+
+  The client will transparently handle re-connection to your Kafka cluster.
+  """
 
-      alias BrodEx.Config
-      alias Brodex.KafkaMessage
+  @type t :: module
 
-      @type endpoint :: String.t
-      @type endpoints :: [endpoint] | String.t
+  @doc false
+  defmacro __using__(use_opts \\ []) do
+    quote bind_quoted: [use_opts: use_opts] do
+      @behaviour BrodEx.Client
+      @otp_app Keyword.fetch!(use_opts, :otp_app)
 
       def start_link(opts \\ []) do
-        config  = Application.get_env(@otp_app, __MODULE__, [])
-        Config.build_config(config)
-        :brod.start(:normal, nil)
+        client_config = Application.get_env(@otp_app, __MODULE__, [])
+                        |> Keyword.merge(opts)
+                        |> BrodEx.Config.brod_client_config()
+        clients = Application.get_env(:brod, :clients, [])
+                  |> Keyword.merge([{__MODULE__, client_config}])
+
+        Application.put_env(:brod, :clients, clients)
+
+        :brod.start_link_client(client_config[:endpoints], __MODULE__, client_config)
```

Comment:
I have a doubt, here you are attaching a client to the brodex supervision tree but in the previous line you add the same client to brod configuration. So in that case if brod supervisor restart for some reason, brod will try to start the same client runing on brodex, doesn't it?

Reply:
Valid point, sorry I think I left that in accidentally. I will remove it.
So my intention is that the user of the library can have its own client under his own supervision.

What do you think of providing a `BrodEx.DefaultClient` within a `BrodEx` supervision tree?

Reply:
I like your idea of have clients under `BrodEx` supervision tree. But we need to support start multiple clients via configuration and `brod` already allow that. That's the reason why i only use `BrodEx` as proxy for `brod` and if the user need a custom set up he always can use the `BrodEx.start_link_client/3`. If we want to move forward with your approach we need to re implement that logic on `BrodEx`.

Reply:
I see. What I found helpful when using Elixir is, for example, that I can setup configuration via environment variables. There are other cases, where I found setting up a client programmatically was very helpful (for applications deployed to Heroku).

So I would like to implement support for a little bit more than `config :brod, :clients, ...`.

I think it would be worth implementing something like default client configuration and a `BrodEx` application.

I see your concern when changing the supervision root to be `BrodEx`, because the current implementation is a good wrapper/proxy already.

What would you suggest to wrap client configuration, and allow for environment variable based configuration? The brod endpoint `{'host', 9092}` tuple is quite annoying to use.